<a href="https://colab.research.google.com/github/Rakesh-SSN/FYP/blob/master/SOFTMAX.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import drive
drive.mount('/content/drive',force_remount=True)

Mounted at /content/drive


The below code reads the output file and converts into a 2 D list of probabilities


In [0]:
def conv(filename,algorithm):
  if(algorithm=="BERT"):
    path="/content/drive/My Drive/FYP/bert/glue/cola/"+filename+".txt"
  elif(algorithm=="USE"):
    path="/content/drive/My Drive/FYP/bert/glue/cola/buffer/USE/"+filename+".txt"
  elif(algorithm=="SEQ"):
    path="/content/drive/My Drive/FYP/bert/glue/cola/buffer/SEQ/"+filename+".txt"
  
  with open(path) as f:
    lang_task = []
    for line in f:
      line=line.replace("[[","").replace("]\n","").replace("]]","").replace("[","").replace("]","")
      line=line.rstrip().lstrip().split(" ")
      #print(line)
      while(line.count('')!=0):
        line.remove('')  
      for i in range(len(line)):
          line[i]=float(line[i])
      lang_task.append(line)
  return lang_task









In [0]:
BERT_task1tamil_results=conv("task1tamil-results","BERT")
BERT_task1hindi_results=conv("task1hindi-results","BERT")
BERT_task1malayalam_results=conv("task1malayalam-results","BERT")
BERT_task1punjabi_results=conv("task1punjabi-results","BERT")


BERT_task2tamil_results=conv("task2tamil-results","BERT")
BERT_task2hindi_results=conv("task2hindi-results","BERT")
BERT_task2malayalam_results=conv("task2malayalam-results","BERT")
BERT_task2punjabi_results=conv("task2punjabi-results","BERT")

USE_task1tamil_results=conv("USE_task1tamil-results","USE")
USE_task1hindi_results=conv("USE_task1hindi-results","USE")
USE_task1malayalam_results=conv("USE_task1malayalam-results","USE")
USE_task1punjabi_results=conv("USE_task1punjabi-results","USE")


USE_task2tamil_results=conv("USE_task2tamil-results","USE")
USE_task2hindi_results=conv("USE_task2hindi-results","USE")
USE_task2malayalam_results=conv("USE_task2malayalam-results","USE")
USE_task2punjabi_results=conv("USE_task2punjabi-results","USE")


SEQ_task1tamil_results=conv("SEQ_task1tamil-results","SEQ")
SEQ_task1hindi_results=conv("SEQ_task1hindi-results","SEQ")
SEQ_task1malayalam_results=conv("SEQ_task1malayalam-results","SEQ")
SEQ_task1punjabi_results=conv("SEQ_task1punjabi-results","SEQ")


SEQ_task2tamil_results=conv("SEQ_task2tamil-results","SEQ")
SEQ_task2hindi_results=conv("SEQ_task2hindi-results","SEQ")
SEQ_task2malayalam_results=conv("SEQ_task2malayalam-results","SEQ")
SEQ_task2punjabi_results=conv("SEQ_task2punjabi-results","SEQ")


In [0]:
def merge(list1,list2,list3):
  result=[]
  for i in range(len(list1)):
    buffer=[]
    for j in range(len(list1[i])):
      buffer.append(list1[i][j])
    for j in range(len(list2[i])):
      buffer.append(list2[i][j])

    result.append(buffer)
    # print(result[i])
  return result

In [0]:
# task1tamil_merge=merge(BERT_task1tamil_results,USE_task1tamil_results,SEQ_task1tamil_results)
# task2tamil_merge=merge(BERT_task2tamil_results,USE_task2tamil_results,SEQ_task2tamil_results)

# task1hindi_merge=merge(BERT_task1hindi_results,USE_task1hindi_results,SEQ_task1hindi_results)
# task2hindi_merge=merge(BERT_task2hindi_results,USE_task2hindi_results,SEQ_task2hindi_results)

# task1malayalam_merge=merge(BERT_task1malayalam_results,USE_task1malayalam_results,SEQ_task1malayalam_results)
# task2malayalam_merge=merge(BERT_task2malayalam_results,USE_task2malayalam_results,SEQ_task2malayalam_results)

# task1punjabi_merge=merge(BERT_task1punjabi_results,USE_task1punjabi_results,SEQ_task1punjabi_results)
# task2punjabi_merge=merge(BERT_task2punjabi_results,USE_task2punjabi_results,SEQ_task2punjabi_results)

In [0]:
# task1tamil_merge=merge(BERT_task1tamil_results,USE_task1tamil_results)
# task2tamil_merge=merge(BERT_task2tamil_results,USE_task2tamil_results)

# task1hindi_merge=merge(BERT_task1hindi_results,USE_task1hindi_results)
# task2hindi_merge=merge(BERT_task2hindi_results,USE_task2hindi_results)

# task1malayalam_merge=merge(BERT_task1malayalam_results,USE_task1malayalam_results)
# task2malayalam_merge=merge(BERT_task2malayalam_results,USE_task2malayalam_results)

# task1punjabi_merge=merge(BERT_task1punjabi_results,USE_task1punjabi_results)
# task2punjabi_merge=merge(BERT_task2punjabi_results,USE_task2punjabi_results)

In [0]:
import numpy as np
def softmax(x):
    """Compute softmax values for each sets of scores in x."""
    #print(len(x))
    e_x = np.exp(x - np.max(x))
    #print(np.max(x))
    return e_x / e_x.sum()

In [0]:
def predict(task_lang_merge,task_number):
  task_lang_results=[]
  line_no=1
  for i in range(len(task_lang_merge)):
    task_lang_results.append(task_lang_merge[i])
  bert_and_use=[]
  for i in range(len(task_lang_merge)):
    print(line_no," ",task_lang_results[i],"--------",end="")
    # max_probability=max(task_lang_results[i])
    x=task_lang_results[i].index(max_probability)
    print(task_lang_results[i].index(max_probability),end="")
    # print("----",x,"---")
    if(task_number==1):
      if(x==0):
        # print("BERT-NP")
        bert_and_use.append(0)
      elif(x==1):
        # print("BERT-P")
        bert_and_use.append(1)
      elif(x==2):
        # print("USE-NP")
        bert_and_use.append(0)
      elif(x==3):
        # print("USE-P")
        bert_and_use.append(1)
    else:
      if(x==0):
        #print("BERT-NP")
        bert_and_use.append(0)
      elif(x==1):
        #print("BERT-P")
        bert_and_use.append(1)
      elif(x==2):
        #print("BERT-SP")
        bert_and_use.append(2)
      elif(x==3):
        #print("USE-NP")
        bert_and_use.append(0)
      elif(x==4):
        #print("USE-P")
        bert_and_use.append(1)
      elif(x==5):
        #print("USE-SP")
        bert_and_use.append(2)
        
    line_no+=1
  return bert_and_use

In [24]:
USE_task1tamil_output=predict(USE_task1tamil_results,1)
USE_task2tamil_output=predict(USE_task2tamil_results,2)

USE_task1hindi_output=predict(USE_task1hindi_results,1)
USE_task2hindi_output=predict(USE_task2hindi_results,2)

USE_task1malayalam_output=predict(USE_task1malayalam_results,1)
USE_task2malayalam_output=predict(USE_task2malayalam_results,2)

USE_task1punjabi_output=predict(USE_task1punjabi_results,1)
USE_task2punjabi_output=predict(USE_task2punjabi_results,2)

Streaming output truncated to the last 5000 lines.
852   [0.68833476, 0.31166524] --------0---- 0 ---
853   [0.7327607, 0.26723933] --------0---- 0 ---
854   [0.68874997, 0.31125006] --------0---- 0 ---
855   [0.66169626, 0.33830371] --------0---- 0 ---
856   [0.7373702, 0.26262978] --------0---- 0 ---
857   [0.55650896, 0.44349104] --------0---- 0 ---
858   [0.6526967, 0.34730324] --------0---- 0 ---
859   [0.6647194, 0.3352806] --------0---- 0 ---
860   [0.6758712, 0.3241288] --------0---- 0 ---
861   [0.67632604, 0.32367393] --------0---- 0 ---
862   [0.6990926, 0.30090737] --------0---- 0 ---
863   [0.5536426, 0.44635752] --------0---- 0 ---
864   [0.5196155, 0.48038453] --------0---- 0 ---
865   [0.4994304, 0.5005696] --------1---- 1 ---
866   [0.71215236, 0.28784764] --------0---- 0 ---
867   [0.6518761, 0.34812394] --------0---- 0 ---
868   [0.70092386, 0.29907608] --------0---- 0 ---
869   [0.5445905, 0.45540953] --------0---- 0 ---
870   [0.6930952, 0.3069048] --------0---- 0 

In [25]:
BERT_task1tamil_output=predict(BERT_task1tamil_results,1)
BERT_task2tamil_output=predict(BERT_task2tamil_results,2)

BERT_task1hindi_output=predict(BERT_task1hindi_results,1)
BERT_task2hindi_output=predict(BERT_task2hindi_results,2)

BERT_task1malayalam_output=predict(BERT_task1malayalam_results,1)
BERT_task2malayalam_output=predict(BERT_task2malayalam_results,2)

BERT_task1punjabi_output=predict(BERT_task1punjabi_results,1)
BERT_task2punjabi_output=predict(BERT_task2punjabi_results,2)

Streaming output truncated to the last 5000 lines.
852   [0.8081368, 0.19186321] --------0---- 0 ---
853   [0.9274097, 0.07259028] --------0---- 0 ---
854   [0.02388916, 0.9761109] --------1---- 1 ---
855   [0.9353757, 0.06462429] --------0---- 0 ---
856   [0.58446205, 0.415538] --------0---- 0 ---
857   [0.28408843, 0.71591157] --------1---- 1 ---
858   [0.72170085, 0.27829918] --------0---- 0 ---
859   [0.9292833, 0.07071669] --------0---- 0 ---
860   [0.9292833, 0.07071669] --------0---- 0 ---
861   [0.9141504, 0.08584954] --------0---- 0 ---
862   [0.9141504, 0.08584954] --------0---- 0 ---
863   [0.4983715, 0.5016285] --------1---- 1 ---
864   [0.8941424, 0.10585763] --------0---- 0 ---
865   [0.02292535, 0.9770747] --------1---- 1 ---
866   [0.5290255, 0.47097448] --------0---- 0 ---
867   [0.94000936, 0.05999056] --------0---- 0 ---
868   [0.85042477, 0.14957523] --------0---- 0 ---
869   [0.93909836, 0.06090162] --------0---- 0 ---
870   [0.9235675, 0.07643259] --------0---- 0 

In [26]:
SEQ_task1tamil_output=predict(SEQ_task1tamil_results,1)
SEQ_task2tamil_output=predict(SEQ_task2tamil_results,2)

SEQ_task1hindi_output=predict(SEQ_task1hindi_results,1)
SEQ_task2hindi_output=predict(SEQ_task2hindi_results,2)

SEQ_task1malayalam_output=predict(SEQ_task1malayalam_results,1)
SEQ_task2malayalam_output=predict(SEQ_task2malayalam_results,2)

SEQ_task1punjabi_output=predict(SEQ_task1punjabi_results,1)
SEQ_task2punjabi_output=predict(SEQ_task2punjabi_results,2)

Streaming output truncated to the last 5000 lines.
852   [0.6, 0.4] --------0---- 0 ---
853   [0.6, 0.4] --------0---- 0 ---
854   [0.6, 0.4] --------0---- 0 ---
855   [0.6, 0.4] --------0---- 0 ---
856   [0.6, 0.4] --------0---- 0 ---
857   [0.6, 0.4] --------0---- 0 ---
858   [0.6, 0.4] --------0---- 0 ---
859   [0.6, 0.4] --------0---- 0 ---
860   [0.6, 0.4] --------0---- 0 ---
861   [0.6, 0.4] --------0---- 0 ---
862   [0.6, 0.4] --------0---- 0 ---
863   [0.6, 0.4] --------0---- 0 ---
864   [0.6, 0.4] --------0---- 0 ---
865   [0.6, 0.4] --------0---- 0 ---
866   [0.6, 0.4] --------0---- 0 ---
867   [0.6, 0.4] --------0---- 0 ---
868   [0.6, 0.4] --------0---- 0 ---
869   [0.6, 0.4] --------0---- 0 ---
870   [0.6, 0.4] --------0---- 0 ---
871   [0.6, 0.4] --------0---- 0 ---
872   [0.6, 0.4] --------0---- 0 ---
873   [0.6, 0.4] --------0---- 0 ---
874   [0.6, 0.4] --------0---- 0 ---
875   [0.6, 0.4] --------0---- 0 ---
876   [0.6, 0.4] --------0---- 0 ---
877   [0.6, 0.4] -------

In [0]:
# task1tamil_output=predict(task1tamil_merge,1)
# task2tamil_output=predict(task2tamil_merge,2)

In [0]:
# task1malayalam_output=predict(task1malayalam_merge,1)
# task2malayalam_output=predict(task2malayalam_merge,2)

In [0]:
# task1hindi_output=predict(task1hindi_merge,1)
# task2hindi_output=predict(task2hindi_merge,2)

In [0]:
# task1punjabi_output=predict(task1punjabi_merge,1)
# task2punjabi_output=predict(task2punjabi_merge,2)

In [0]:
def get_truelabels(filename):
  
  path="/content/drive/My Drive/FYP/bert/glue/cola/buffer/true labels/"+filename+".txt"
  with open(path) as f:
    truelabels = []
    for line in f:
      line=line.replace("[","").replace("]\n","").replace("]","")
      line=line.rstrip().lstrip().split(" ")  
      for i in range(len(line)):
        line[i]=int(line[i])
        truelabels.append(line[i])
  return truelabels

task1tamil_truelabels=get_truelabels("task1tamil-labels")
task1hindi_truelabels=get_truelabels("task1hindi-labels")
task1malayalam_truelabels=get_truelabels("task1malayalam-labels")
task1punjabi_truelabels=get_truelabels("task1punjabi-labels")
task2tamil_truelabels=get_truelabels("task2tamil-labels")
task2hindi_truelabels=get_truelabels("task2hindi-labels")
task2malayalam_truelabels=get_truelabels("task2malayalam-labels")
task2punjabi_truelabels=get_truelabels("task2punjabi-labels")

In [0]:
def eval_predictions(task_lang_output,task_lang_truelabels):
  true_predictions=0
  false_predictions=0
  for i in range(len(task_lang_output)):
    if(task_lang_output[i]==task_lang_truelabels[i]):
      true_predictions+=1
    else:
      false_predictions+=1


  print("\nTrue predictions:",true_predictions)
  print("False predictions:",false_predictions)
  
  # print(task_lang_output)
  # print(task_lang_truelabels)
  accuracy=true_predictions/len(task_lang_output)*100
  print("Acuuracy:",accuracy,"%")
  print("**********************************************")



In [0]:
# print("Task1-Tamil:",end="")
# task1tamil_eval=eval_predictions(task1tamil_output,task1tamil_truelabels)
# print("Task2-Tamil:",end="")
# task2tamil_eval=eval_predictions(task2tamil_output,task2tamil_truelabels)
# print("----------------------------------------------------------------------------------------------------------------")
# print("Task1-Hindi:",end="")
# task1hindi_eval=eval_predictions(task1hindi_output,task1hindi_truelabels)
# print("Task2-Hindi:",end="")
# task2hindi_eval=eval_predictions(task2hindi_output,task2hindi_truelabels)
# print("----------------------------------------------------------------------------------------------------------------")
# print("Task1-Malayalam:",end="")
# task1malayalam_eval=eval_predictions(task1malayalam_output,task1malayalam_truelabels)
# print("Task2-Malayalam:",end="")
# task2malayalam_eval=eval_predictions(task2malayalam_output,task2malayalam_truelabels)
# print("----------------------------------------------------------------------------------------------------------------")
# print("Task1-Punjabi:",end="")
# task1punjabi_eval=eval_predictions(task1punjabi_output,task1punjabi_truelabels)
# print("Task2-Punjabi:",end="")
# task2punjabi_eval=eval_predictions(task2punjabi_output,task2punjabi_truelabels)
# print("----------------------------------------------------------------------------------------------------------------")

In [0]:
from sklearn.metrics import precision_score
from sklearn.metrics import f1_score

**BERT**

---



In [45]:
print("BERT RESULTS")
print("----------------------------------------------------------------------------------------------------------------")
print("Task1-Tamil:",end="")
BERT_task1tamil_eval=eval_predictions(BERT_task1tamil_output,task1tamil_truelabels)
print(precision_score(BERT_task1tamil_output,task1tamil_truelabels))

print("Task2-Tamil:",end="")
BERT_task2tamil_eval=eval_predictions(BERT_task2tamil_output,task2tamil_truelabels)
print("----------------------------------------------------------------------------------------------------------------")
print("Task1-Hindi:",end="")
BERT_task1hindi_eval=eval_predictions(BERT_task1hindi_output,task1hindi_truelabels)
print("Task2-Hindi:",end="")
BERT_task2hindi_eval=eval_predictions(BERT_task2hindi_output,task2hindi_truelabels)
print("----------------------------------------------------------------------------------------------------------------")
print("Task1-Malayalam:",end="")
BERT_task1malayalam_eval=eval_predictions(BERT_task1malayalam_output,task1malayalam_truelabels)
print("Task2-Malayalam:",end="")
BERT_task2malayalam_eval=eval_predictions(BERT_task2malayalam_output,task2malayalam_truelabels)
print("----------------------------------------------------------------------------------------------------------------")
print("Task1-Punjabi:",end="")
BERT_task1punjabi_eval=eval_predictions(BERT_task1punjabi_output,task1punjabi_truelabels)
print("Task2-Punjabi:",end="")
BERT_task2punjabi_eval=eval_predictions(BERT_task2punjabi_output,task2punjabi_truelabels)
print("----------------------------------------------------------------------------------------------------------------")

BERT RESULTS
----------------------------------------------------------------------------------------------------------------
Task1-Tamil:
True predictions: 526
False predictions: 374
Acuuracy: 58.44444444444444 %
**********************************************
0.2475
Task2-Tamil:
True predictions: 629
False predictions: 771
Acuuracy: 44.92857142857143 %
**********************************************
----------------------------------------------------------------------------------------------------------------
Task1-Hindi:
True predictions: 765
False predictions: 135
Acuuracy: 85.0 %
**********************************************
Task2-Hindi:
True predictions: 766
False predictions: 634
Acuuracy: 54.714285714285715 %
**********************************************
----------------------------------------------------------------------------------------------------------------
Task1-Malayalam:
True predictions: 589
False predictions: 311
Acuuracy: 65.44444444444444 %
*********************

**USE**

In [30]:
print("USE RESULTS")
print("----------------------------------------------------------------------------------------------------------------")
print("Task1-Tamil:",end="")
USE_task1tamil_eval=eval_predictions(USE_task1tamil_output,task1tamil_truelabels)
print("Task2-Tamil:",end="")
USE_task2tamil_eval=eval_predictions(USE_task2tamil_output,task2tamil_truelabels)
print("----------------------------------------------------------------------------------------------------------------")
print("Task1-Hindi:",end="")
USE_task1hindi_eval=eval_predictions(USE_task1hindi_output,task1hindi_truelabels)
print("Task2-Hindi:",end="")
USE_task2hindi_eval=eval_predictions(USE_task2hindi_output,task2hindi_truelabels)
print("----------------------------------------------------------------------------------------------------------------")
print("Task1-Malayalam:",end="")
USE_task1malayalam_eval=eval_predictions(USE_task1malayalam_output,task1malayalam_truelabels)
print("Task2-Malayalam:",end="")
USE_task2malayalam_eval=eval_predictions(USE_task2malayalam_output,task2malayalam_truelabels)
print("----------------------------------------------------------------------------------------------------------------")
print("Task1-Punjabi:",end="")
USE_task1punjabi_eval=eval_predictions(USE_task1punjabi_output,task1punjabi_truelabels)
print("Task2-Punjabi:",end="")
USE_task2punjabi_eval=eval_predictions(USE_task2punjabi_output,task2punjabi_truelabels)
print("----------------------------------------------------------------------------------------------------------------")

USE RESULTS
----------------------------------------------------------------------------------------------------------------
Task1-Tamil:
True predictions: 486
False predictions: 414
Acuuracy: 54.0 %
**********************************************
Task2-Tamil:
True predictions: 308
False predictions: 1092
Acuuracy: 22.0 %
**********************************************
----------------------------------------------------------------------------------------------------------------
Task1-Hindi:
True predictions: 504
False predictions: 396
Acuuracy: 56.00000000000001 %
**********************************************
Task2-Hindi:
True predictions: 482
False predictions: 918
Acuuracy: 34.42857142857143 %
**********************************************
----------------------------------------------------------------------------------------------------------------
Task1-Malayalam:
True predictions: 505
False predictions: 395
Acuuracy: 56.111111111111114 %
*****************************************

In [31]:
print("Seq2Seq RESULTS")
print("----------------------------------------------------------------------------------------------------------------")
print("Task1-Tamil:",end="")
SEQ_task1tamil_eval=eval_predictions(SEQ_task1tamil_output,task1tamil_truelabels)
print("Task2-Tamil:",end="")
SEQ_task2tamil_eval=eval_predictions(SEQ_task2tamil_output,task2tamil_truelabels)
print("----------------------------------------------------------------------------------------------------------------")
print("Task1-Hindi:",end="")
SEQ_task1hindi_eval=eval_predictions(SEQ_task1hindi_output,task1hindi_truelabels)
print("Task2-Hindi:",end="")
SEQ_task2hindi_eval=eval_predictions(SEQ_task2hindi_output,task2hindi_truelabels)
print("----------------------------------------------------------------------------------------------------------------")
print("Task1-Malayalam:",end="")
SEQ_task1malayalam_eval=eval_predictions(SEQ_task1malayalam_output,task1malayalam_truelabels)
print("Task2-Malayalam:",end="")
SEQ_task2malayalam_eval=eval_predictions(SEQ_task2malayalam_output,task2malayalam_truelabels)
print("----------------------------------------------------------------------------------------------------------------")
print("Task1-Punjabi:",end="")
SEQ_task1punjabi_eval=eval_predictions(SEQ_task1punjabi_output,task1punjabi_truelabels)
print("Task2-Punjabi:",end="")
SEQ_task2punjabi_eval=eval_predictions(SEQ_task2punjabi_output,task2punjabi_truelabels)
print("----------------------------------------------------------------------------------------------------------------")

Seq2Seq RESULTS
----------------------------------------------------------------------------------------------------------------
Task1-Tamil:
True predictions: 474
False predictions: 426
Acuuracy: 52.666666666666664 %
**********************************************
Task2-Tamil:
True predictions: 694
False predictions: 706
Acuuracy: 49.57142857142857 %
**********************************************
----------------------------------------------------------------------------------------------------------------
Task1-Hindi:
True predictions: 500
False predictions: 400
Acuuracy: 55.55555555555556 %
**********************************************
Task2-Hindi:
True predictions: 476
False predictions: 924
Acuuracy: 34.0 %
**********************************************
----------------------------------------------------------------------------------------------------------------
Task1-Malayalam:
True predictions: 522
False predictions: 378
Acuuracy: 57.99999999999999 %
*************************

In [0]:
# from sklearn.metrics import accuracy_score
# accuracy_score(task1tamil_output,task1tamil_truelabels)

In [0]:
# from sklearn.metrics import precision_score
# precision_score(task1tamil_output,task1tamil_truelabels,average='weighted')

In [0]:
# from sklearn.metrics import recall_score
# recall_score(task1tamil_output,task1tamil_truelabels, average='weighted')

In [0]:
# from sklearn.metrics import f1_score
# f1_score(task1tamil_output,task1tamil_truelabels, average='weighted')

In [0]:
def generate_combined_probability_average(bert,use,seq,task):
  overall_probability=[]
  # print(bert)
  # print(use)
  # print(seq)
  if(task==1):
    for i in range(len(bert)):
      average_NP=(bert[i][0]+use[i][0]+seq[i][0])/3
      average_P=(bert[i][1]+use[i][1]+seq[i][1])/3
      overall_probability.append([average_NP,average_P])
  else:
    for i in range(len(bert)):
      average_NP=(bert[i][0]+use[i][0]+seq[i][0])/3
      average_P=(bert[i][1]+use[i][1]+seq[i][1])/3
      average_SP=(bert[i][2]+use[i][2]+seq[i][2])/3
      overall_probability.append([average_NP,average_P,average_SP])

  return overall_probability


def generate_combined_probability_majority(bert,use,seq,task,all_results):

  BERT_current_lang_task_prob=all_results[0]
  USE_current_lang_task_prob=all_results[1]
  SEQ_current_lang_task_prob=all_results[2]

  overall_prediction=[]
  if(task==1):
    for i in range(len(bert)):
      row=[bert[i],use[i],seq[i]]
      NP=row.count(0)
      P=row.count(1)
      if(NP>P):
        overall_prediction.append(0)
      elif(P>NP):
        overall_prediction.append(1)
  else:
    for i in range(len(bert)):
      row=[bert[i],use[i],seq[i]]
      NP=row.count(0)
      P=row.count(1)
      SP=row.count(2)
      if(NP>=2):
        overall_prediction.append(0)
      elif(P>=2):
        overall_prediction.append(1)
      elif(SP>=2):
        overall_prediction.append(2)
      else:
        bert_prob_of_pred=BERT_current_lang_task_prob[i][bert[i]]
        use_prob_of_pred=USE_current_lang_task_prob[i][use[i]]
        seq_prob_of_pred=SEQ_current_lang_task_prob[i][seq[i]]
    
        max_prob_for_predictions=max(bert_prob_of_pred,
                                    use_prob_of_pred,
                                    seq_prob_of_pred)
        if(max_prob_for_predictions==bert_prob_of_pred):
          overall_prediction.append(bert[i])
        elif(max_prob_for_predictions==use_prob_of_pred):
          overall_prediction.append(use[i])
        elif(max_prob_for_predictions==seq_prob_of_pred):
          overall_prediction.append(seq[i])

  return overall_prediction

def predict_results(result_list):
  final_result=[]
  for i in range(len(result_list)):
    pos=result_list[i].index(max(result_list[i]))
    final_result.append(pos)
  return final_result




      else:
        bert_prob_of_pred=BERT_current_lang_task_prob[i][bert[i]]
        use_prob_of_pred=USE_current_lang_task_prob[i][use[i]]
        seq_prob_of_pred=SEQ_current_lang_task_prob[i][seq[i]]
        
        max_prob_for_predictions=max(bert_prob_of_pred,
                                     use_prob_of_pred,
                                     seq_prob_of_pred)
        if(max_prob_for_predictions==bert_prob_of_pred):
          overall_prediction.append(bert[i])
        elif(max_prob_for_predictions==use_prob_of_pred):
          overall_prediction.append(use[i])
        elif(max_prob_for_predictions==seq_prob_of_pred):
          overall_prediction.append(seq[i])

In [0]:
# task1tamil_combined_probability=generate_combined_probability_average(BERT_task1tamil_results,USE_task1tamil_results,SEQ_task1tamil_results,1)
# task1hindi_combined_probability=generate_combined_probability_average(BERT_task1hindi_results,USE_task1hindi_results,SEQ_task1hindi_results,1)
# task1malayalam_combined_probability=generate_combined_probability_average(BERT_task1malayalam_results,USE_task1malayalam_results,SEQ_task1malayalam_results,1)
# task1punjabi_combined_probability=generate_combined_probability_average(BERT_task1punjabi_results,USE_task1punjabi_results,SEQ_task1punjabi_results,1)

# print("TASK1-TAMIL")
# task1tamil_combined_predictions=predict_results(task1tamil_combined_probability)
# print(task1tamil_combined_predictions)
# task1tamil_combined_eval=eval_predictions(task1tamil_combined_predictions,task1tamil_truelabels)
# print("TASK1-HINDI")
# task1hindi_combined_predictions=predict_results(task1hindi_combined_probability)
# print(task1hindi_combined_predictions)
# task1hindi_combined_eval=eval_predictions(task1hindi_combined_predictions,task1hindi_truelabels)
# print("TASK1-MALAYALAM")
# task1malayalam_combined_predictions=predict_results(task1malayalam_combined_probability)
# print(task1tamil_combined_predictions)
# task1malayalam_combined_eval=eval_predictions(task1malayalam_combined_predictions,task1malayalam_truelabels)
# print("TASK1-PUNJABI")
# task1punjabi_combined_predictions=predict_results(task1punjabi_combined_probability)
# print(task1punjabi_combined_predictions)
# task1punjabi_combined_eval=eval_predictions(task1punjabi_combined_predictions,task1punjabi_truelabels)

In [0]:
# task2tamil_combined_probability=generate_combined_probability_average(BERT_task2tamil_results,USE_task2tamil_results,SEQ_task2tamil_results,2)
# task2hindi_combined_probability=generate_combined_probability_average(BERT_task2hindi_results,USE_task2hindi_results,SEQ_task2hindi_results,2)
# task2malayalam_combined_probability=generate_combined_probability_average(BERT_task2malayalam_results,USE_task2malayalam_results,SEQ_task2malayalam_results,2)
# task2punjabi_combined_probability=generate_combined_probability_average(BERT_task2punjabi_results,USE_task2punjabi_results,SEQ_task2punjabi_results,2)

# print("TASK2-TAMIL")
# task2tamil_combined_predictions=predict_results(task2tamil_combined_probability)
# #print(task2tamil_combined_predictions)
# task2tamil_combined_eval=eval_predictions(task2tamil_combined_predictions,task2tamil_truelabels)
# print("TASK2-HINDI")
# task2hindi_combined_predictions=predict_results(task2hindi_combined_probability)
# #print(task2hindi_combined_predictions)
# task2hindi_combined_eval=eval_predictions(task2hindi_combined_predictions,task2hindi_truelabels)
# print("TASK2-MALAYALAM")
# task2malayalam_combined_predictions=predict_results(task2malayalam_combined_probability)
# #print(task2malayalam_combined_predictions)
# task2malayalam_combined_eval=eval_predictions(task2malayalam_combined_predictions,task2malayalam_truelabels)
# print("TASK2-PUNJABI")
# task2punjabi_combined_predictions=predict_results(task2punjabi_combined_probability)
# #print(task2punjabi_combined_predictions)
# task2punjabi_combined_eval=eval_predictions(task2punjabi_combined_predictions,task2punjabi_truelabels)

In [40]:
task1tamil_probabilities=[BERT_task1tamil_results,USE_task1tamil_results,SEQ_task1tamil_results]
task1tamil_majority=generate_combined_probability_majority(BERT_task1tamil_output,USE_task1tamil_output,SEQ_task1tamil_output,1,task1tamil_probabilities)

task1hindi_probabilities=[BERT_task1hindi_results,USE_task1hindi_results,SEQ_task1hindi_results]
task1hindi_majority=generate_combined_probability_majority(BERT_task1hindi_output,USE_task1hindi_output,SEQ_task1hindi_output,1,task1hindi_probabilities)

task1malayalam_probabilities=[BERT_task1malayalam_results,USE_task1malayalam_results,SEQ_task1malayalam_results]
task1malayalam_majority=generate_combined_probability_majority(BERT_task1malayalam_output,USE_task1malayalam_output,SEQ_task1malayalam_output,1,task1malayalam_probabilities)

task1punjabi_probabilities=[BERT_task1punjabi_results,USE_task1punjabi_results,SEQ_task1punjabi_results]
task1punjabi_majority=generate_combined_probability_majority(BERT_task1punjabi_output,USE_task1punjabi_output,SEQ_task1punjabi_output,1,task1punjabi_probabilities)
print("True labels vs Predicted labels")
print(task1tamil_truelabels)
print(task1tamil_majority)
task1tamil_majority_eval=eval_predictions(task1tamil_majority,task1tamil_truelabels)

print("True labels vs Predicted labels")
print(task1hindi
      _truelabels)
print((task1hindi_majority))
task1hindi_majority_eval=eval_predictions(task1hindi_majority,task1hindi_truelabels)
print("True labels vs Predicted labels")
print(task1malayalam_truelabels)
print(task1malayalam_majority)
task1malayalam_majority_eval=eval_predictions(task1malayalam_majority,task1malayalam_truelabels)
print("True labels vs Predicted labels")
print(task1punjabi_truelabels)
print(task1punjabi_majority)
task1punjabi_majority_eval=eval_predictions(task1punjabi_majority,task1punjabi_truelabels)

True labels vs Predicted labels
[1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0


True predictions: 500
False predictions: 400
Acuuracy: 55.55555555555556 %
**********************************************


True predictions: 511
False predictions: 389
Acuuracy: 56.777777777777786 %
**********************************************


True predictions: 541
False predictions: 359
Acuuracy: 60.111111111111114 %
**********************************************

True predictions: 301
False predictions: 199
Acuuracy: 60.199999999999996 %
**********************************************


In [42]:
task2tamil_probabilities=[BERT_task2tamil_results,USE_task2tamil_results,SEQ_task2tamil_results]
task2tamil_majority=generate_combined_probability_majority(BERT_task2tamil_output,USE_task2tamil_output,SEQ_task2tamil_output,2,task2tamil_probabilities)

task2hindi_probabilities=[BERT_task2hindi_results,USE_task2hindi_results,SEQ_task2hindi_results]
task2hindi_majority=generate_combined_probability_majority(BERT_task2hindi_output,USE_task2hindi_output,SEQ_task2hindi_output,2,task2hindi_probabilities)

task2malayalam_probabilities=[BERT_task2malayalam_results,USE_task2malayalam_results,SEQ_task2malayalam_results]
task2malayalam_majority=generate_combined_probability_majority(BERT_task2malayalam_output,USE_task2malayalam_output,SEQ_task2malayalam_output,2,task2malayalam_probabilities)

task2punjabi_probabilities=[BERT_task2punjabi_results,USE_task2punjabi_results,SEQ_task2punjabi_results]
task2punjabi_majority=generate_combined_probability_majority(BERT_task2punjabi_output,USE_task2punjabi_output,SEQ_task2punjabi_output,2,task2punjabi_probabilities)
print("TASK2TAMIL-True labels vs Predicted labels")
print(task2tamil_truelabels)
print(task2tamil_majority)
task2tamil_majority_eval=eval_predictions(task2tamil_majority,task2tamil_truelabels)
print("TASK2HINDI-True labels vs Predicted labels")
print(task2hindi_truelabels)
print((task2hindi_majority))
task2hindi_majority_eval=eval_predictions(task2hindi_majority,task2hindi_truelabels)
print("TASK2MALAYALAM-True labels vs Predicted labels")
print(task2malayalam_truelabels)
print(task2malayalam_majority)
task2malayalam_majority_eval=eval_predictions(task2malayalam_majority,task2malayalam_truelabels)
print("TASK2PUNJABI-True labels vs Predicted labels")
print(task2punjabi_truelabels)
print(task2punjabi_majority)
task2punjabi_majority_eval=eval_predictions(task2punjabi_majority,task2punjabi_truelabels)

TASK2TAMIL-True labels vs Predicted labels
[0, 2, 2, 0, 2, 0, 2, 2, 1, 2, 1, 0, 0, 0, 2, 0, 0, 0, 0, 2, 0, 0, 2, 2, 1, 2, 1, 2, 2, 2, 0, 2, 2, 0, 1, 1, 0, 0, 0, 1, 1, 2, 2, 2, 2, 1, 2, 1, 2, 1, 1, 1, 1, 2, 0, 1, 0, 2, 0, 1, 1, 1, 1, 1, 0, 0, 2, 2, 2, 2, 0, 0, 2, 0, 2, 2, 1, 2, 0, 2, 0, 1, 1, 2, 0, 2, 1, 1, 1, 1, 0, 2, 1, 1, 0, 1, 2, 0, 0, 0, 1, 1, 1, 2, 2, 2, 1, 2, 0, 0, 1, 1, 0, 2, 2, 2, 1, 1, 1, 1, 1, 2, 2, 0, 1, 1, 0, 0, 2, 0, 0, 0, 2, 1, 2, 1, 2, 0, 0, 0, 0, 2, 1, 2, 1, 0, 2, 2, 0, 2, 0, 0, 1, 0, 0, 2, 0, 0, 2, 0, 0, 2, 1, 0, 2, 0, 0, 0, 2, 0, 1, 1, 2, 0, 0, 0, 2, 2, 2, 0, 0, 1, 1, 0, 0, 2, 0, 0, 0, 2, 1, 1, 1, 2, 1, 1, 1, 2, 2, 2, 2, 0, 0, 0, 2, 2, 2, 2, 2, 2, 2, 0, 0, 0, 0, 0, 0, 1, 2, 2, 0, 2, 0, 1, 0, 0, 2, 0, 2, 0, 2, 0, 2, 2, 0, 0, 0, 1, 2, 0, 0, 2, 0, 0, 2, 2, 0, 0, 2, 0, 1, 1, 0, 0, 1, 1, 0, 2, 1, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 2, 0, 1, 2, 0, 0, 0, 2, 1, 1, 1, 0, 2, 2, 2, 1, 1, 2, 0, 2, 1, 2, 2, 1, 1, 0, 0, 0, 1, 1, 1, 2, 2, 1, 1, 0, 2, 2, 0, 2, 2, 1, 2, 2, 0, 2, 0, 2, 2,